File: 3A/se382/A2/lsystem/TurtleGraphics.java
/*
 *
 */
package lsystem;

import java.awt.geom.AffineTransform;
import java.awt.geom.Point2D;

public class TurtleGraphics {

    private Point2D.Double curLocation = new Point2D.Double();
    private Point2D.Double curAngle    = new Point2D.Double();
    
    public Point2D.Double getLocation() {
        return (Point2D.Double)curLocation.clone();
    }
    
    public void setLocation(Point2D.Double location) {
        curLocation.setLocation(location);
    }
    
    public Point2D.Double getAngleVector() {
        return (Point2D.Double)curAngle.clone();
    }
    public void setAngleVector(Point2D.Double angle) {
        curAngle.setLocation(angle);
    }
    
    public Point2D.Double moveForward() {
        curLocation.x = curLocation.x + curAngle.x;
        curLocation.y = curLocation.y + curAngle.y;
        return this.getLocation();
    }
    
    public void pointLeft() {
        curAngle.setLocation(-1, 0);
    }
    public void pointRight() {
        curAngle.setLocation(1, 0);
    }
    public void pointUp() {
        curAngle.setLocation(0, -1);
    }
    public void pointDown() {
        curAngle.setLocation(0, 1);
    }
    public void turn(double angle) {
        AffineTransform t = new AffineTransform();
        t.rotate(Math.toRadians(angle));
        t.transform(this.curAngle, this.curAngle);
    }
}

File: 3A/se382/A1/lsystem/TurtleGraphics.cpp
#include "TurtleGraphics.h"
#include "../AffineTransform.h"

namespace cs349
{

TurtleGraphics::TurtleGraphics()
{
	this->curLocation     = Point(0,0);
	this->directionVector = Point(1,0);
}

TurtleGraphics::~TurtleGraphics()
{
}

Point TurtleGraphics::GetLocation() const
{
	return this->curLocation;
}

void TurtleGraphics::SetLocation(const Point & p)
{
	this->curLocation = p;
}

Point TurtleGraphics::GetAngleVector() const
{
	return this->directionVector;
}

void  TurtleGraphics::SetAngleVector(const Point & p)
{
	this->directionVector = p;
}


Point TurtleGraphics::MoveForward()
{
	this->curLocation.x += this->directionVector.x;
	this->curLocation.y += this->directionVector.y;
	return this->curLocation;
}

void  TurtleGraphics::PointLeft()
{
	this->directionVector = Point(-1, 0);
}

void  TurtleGraphics::PointRight()
{
	this->directionVector = Point(1, 0);
}

void  TurtleGraphics::PointUp()
{
	this->directionVector = Point(0, -1);
}

void  TurtleGraphics::PointDown()
{
	this->directionVector = Point(0, 1);
}

void  TurtleGraphics::Turn(double angle)
{
	AffineTransform t;
	t.RotateInDegrees(angle);
	this->directionVector = t.Transform(this->directionVector);
}

}
